### ISSUE TITLE
App pod fails to start in KubeCF v.0.2.0 - cannot pull image as os="linux" not set.

### ISSUE THREAD

**Issue by danielvaughan** [cloudfoundry-incubator/kubecf#491]
**Describe the bug**
I have an installation of v0.2.0 deployed to Kubernetes with Eirini enabled.

I cannot deploy an application that creates a new container.

The application builds but the resultant container cannot be started.

This appears to apply to all applications I have attempted that need to build a container. The php example below is representative.

If I enable docker support and do a CF push with an existing container it starts without issue.

**To Reproduce**
I attempt to deploy the cf-ex-php-info example from the cloudfoundry-samples git repo using cf push.
```
git clone https://github.com/cloudfoundry-samples/cf-ex-php-info
cd cf-ex-php-info
cf push --random-route

cf push --random-route
Pushing from manifest to org demo-org / space demo as admin...
Using manifest file /Users/danielvaughan/tmp/cf-ex-php-info/manifest.yml
Getting app info...
Creating app with these attributes...
+ name:         cf-ex-php-info
  path:         /Users/danielvaughan/tmp/cf-ex-php-info
  buildpacks:
+   php_buildpack
+ memory:       128M
  env:
+   SOME_VAR
  routes:
+   cf-ex-php-info-wacky-dingo-pj.169.46.71.67.nip.io

Creating app cf-ex-php-info...
Mapping routes...
Comparing local files to remote cache...
Packaging files to upload...
Uploading files...
 3.21 KiB / 3.21 KiB [=================================================================================================] 100.00% 1s

Waiting for API to complete processing files...

Staging app and tracing logs...
   2020/03/17 09:14:20 executor-started
   -------> Buildpack version 4.4.2
   2020/03/17 09:14:18 downloader-started
   2020/03/17 09:14:18 Installing dependencies
   2020/03/17 09:14:19 downloader-done
   Installing HTTPD
   HTTPD 2.4.41
   Downloaded [https://buildpacks.cloudfoundry.org/dependencies/httpd/httpd-2.4.41-linux-x64-cflinuxfs3-9e9b6146.tgz] to [/tmp]

   Warning: PHP_EXTENSIONS in options.json is deprecated. See: http://docs.cloudfoundry.org/buildpacks/php/gsg-php-config.html
   Installing PHP
   Downloaded [https://buildpacks.cloudfoundry.org/dependencies/php/php7-7.2.24-linux-x64-cflinuxfs3-7a9bcb0f.tgz] to [/tmp]

   Finished: [2020-03-17 09:14:29.392473]
   2020/03/17 09:14:51 executor-done
   2020/03/17 09:14:52 uploader-started
   2020/03/17 09:15:01 uploader-done

Waiting for app to start...
Start unsuccessful

TIP: use 'cf logs cf-ex-php-info --recent' for more information
FAILED
```

The application appears to build successfully but cannot start.

If I check the pod I see:

```
cf-ex-php-info-demo-8248cfe9f0-0                        0/1     ImagePullBackOff   0          4m36s
```

Looking at the events I see:

```
Events:
  Type     Reason     Age                  From                     Message
  ----     ------     ----                 ----                     -------
  Normal   Scheduled  <unknown>            default-scheduler        Successfully assigned kubecf-eirini/cf-ex-php-info-demo-8248cfe9f0-0 to 10.171.110.200
  Normal   BackOff    80s (x5 over 5m17s)  kubelet, 10.171.110.200  Back-off pulling image "127.0.0.1:32123/cloudfoundry/637543ca-9322-42bd-8516-e06788137900:a19b8b0e6c9f2b09742f531005573435caf36e2e"
  Warning  Failed     80s (x5 over 5m17s)  kubelet, 10.171.110.200  Error: ImagePullBackOff
  Normal   Pulling    66s (x4 over 6m15s)  kubelet, 10.171.110.200  Pulling image "127.0.0.1:32123/cloudfoundry/637543ca-9322-42bd-8516-e06788137900:a19b8b0e6c9f2b09742f531005573435caf36e2e"
  Warning  Failed     14s (x4 over 5m17s)  kubelet, 10.171.110.200  Failed to pull image "127.0.0.1:32123/cloudfoundry/637543ca-9322-42bd-8516-e06788137900:a19b8b0e6c9f2b09742f531005573435caf36e2e": rpc error: code = Unknown desc = Image operating system mismatch: image uses "", expecting "linux"
  Warning  Failed     14s (x4 over 5m17s)  kubelet, 10.171.110.200  Error: ErrImagePull
```
It looks like the container is generated by CF and uploaded but the OS not set. This means when the container pull is attempted a suitable "linux:" container is not found so no image can be pulled.

**Expected behavior**
I expected the pod to start successfully in the kubecf-eirini namespace after pulling a container image.

**Environment**
 - CF Version cf version 6.50.0+4f0c3a2ce.2020-03-03
 - Component/Plugin Name
 - Others

**Additional context**

**Comment by danielvaughan** [comment]
This looks very similar to this issue reported with podman builds:
https://github.com/containers/libpod/issues/5503

**Comment by f0rmiga** [comment]
@danielvaughan Thanks for reporting this issue. What's your k8s version?

**Comment by danielvaughan** [comment]
Thanks this is running on OpenShift 4.3.1 so Kubernetes 1.16.2

It looks likely that the check for os="linux" may be OpenShift specific. Does KubeCF or dependencies use podman to build container images as this fixed issue from podman may be applicable: https://github.com/containers/libpod/issues/5503

**Comment by f0rmiga** [comment]
@danielvaughan The KubeCF images are built with Docker via [Fissile](https://github.com/cloudfoundry-incubator/fissile). The issue you are hitting is related to Eirini. I'd suggest reporting the bug in the Eirini repo.

**Comment by danielvaughan** [comment]
Thanks, what version of Eirini is KubeCF v0.2.0 using?

**Comment by danielvaughan** [comment]
It looks like the os not being set should not be an issue. The problem is that on 22nd Nov 2019 an aggressive check for OS was added to copy in containers/image (https://github.com/containers/image/blob/d66f61bca2276786c2362ee33113dbb8f6012299/copy/copy.go). However, this was removed on 31st Jan 2020 due to the problems it caused.

It looks like originally the intention was to check that windows containers were not scheduled on linux and visa-vera but this got checked to a check on an exact match of OS. 

Before this check was added having no os set would not have been a problem. I will look into Eirini and see is this particular version on the copy code is used.

**Comment by f0rmiga** [comment]
I'll close this issue for now. Feel free to re-open in the future if you feel there's any action from our side directly related to this issue.

**Comment by danielvaughan** [comment]
This is related to the particular version of OpenShift we used (4.3.1). The CRI-O container runtime 1v.16.x  in turn, uses v5.2.0 of containers/image. This contains a strict check on os. This is resolved/removed in v5.2.1 but this fix has not made it into CRI-O until 1.17.x currently, so not into OpenShift.

**Comment by danielvaughan** [comment]
No issue for KubeCF but a blocking issue for anyone who wants to get KubeCF running on OpenShift 4.3 currently.
